### ISSUE TITLE
Flakiness in test_paths in CI

### ISSUE THREAD

**Issue by neerajprad** [uber/pyro#1427]
It seems like some of the parametrized tests are failing. @fritzo: Is this expected? If the flop counts are not deterministic, should we test within a broader margin? 

Failing CI build - https://travis-ci.org/uber/pyro/jobs/435351412.

```bash
tests/ops/einsum/test_paths.py:166: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
equation = 'chd,bde,agbc,hiad,hgc,hgi,hiad->'
shapes = [(3, 4, 2), (3, 2, 2), (4, 3, 3, 3), (4, 2, 4, 2), (4, 3, 3), (4, 3, 2), ...]
    def _test_path(equation, shapes):
        inputs, output = equation.split('->')
        inputs = inputs.split(',')
        dim_sizes = {dim: size for dims, shape in zip(inputs, shapes)
                     for dim, size in zip(dims, shape)}
        operands = [torch.randn(shape).log() for shape in shapes]
        logging.debug(equation)
    
        # Compute path using Pyro.
        pyro_time = -timeit.default_timer()
        pyro_path = optimize(inputs, output, dim_sizes)
        _, pyro_info = opt_einsum.contract_path(equation, *operands, path=pyro_path)
        pyro_time += timeit.default_timer()
        pyro_info = '\n'.join(pyro_info.splitlines()[1:7])
    
        assert sum(map(len, pyro_path)) - len(pyro_path) + 1 == len(inputs)
        path = ssa_to_linear(linear_to_ssa(pyro_path))
        assert path == pyro_path
    
        # Compute path using opt_einsum's greedy method.
        opt_time = -timeit.default_timer()
        opt_path, opt_info = opt_einsum.contract_path(equation, *operands, path='greedy')
        opt_time += timeit.default_timer()
        opt_info = '\n'.join(opt_info.splitlines()[1:7])
    
        # Check path quality.
        logging.debug(u'Pyro path took {}s:\n{}'.format(pyro_time, pyro_info))
        logging.debug(u'opt_einsum took {}s:\n{}'.format(opt_time, opt_info))
        pyro_flops = float(re.search('Optimized FLOP count:(.*)', pyro_info).group(1))
        opt_flops = float(re.search('Optimized FLOP count:(.*)', opt_info).group(1))
>       assert pyro_flops <= opt_flops * 1.5 + 2.0
E       assert 3433.0 <= ((2080.0 * 1.5) + 2.0)

```

**Comment by neerajprad** [comment]
I am able to see this locally for at least one of the cases - `bd,db,eac->ace`, when I run it a few times. 

```
FD 	 bd,db,eac->ace
D 	 Pyro path took 0.0013107859995216131s:
         Naive scaling:  5
     Optimized scaling:  3
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  6.000e+01
   Theoretical speedup:  7.200
  Largest intermediate:  2.400e+01 elements
D 	 opt_einsum took 0.0009794710204005241s:
         Naive scaling:  5
     Optimized scaling:  3
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  3.600e+01
   Theoretical speedup:  12.000
  Largest intermediate:  2.400e+01 elements
```

**Comment by fritzo** [comment]
Confirmed I can reproduce the difference in Python 3
```
$ pytest -vs tests/ops/einsum/test_paths.py -k 'bd,db,eac->ace'
==== test session starts ====
platform darwin -- Python 3.5.4, pytest-3.7.0, py-1.6.0, pluggy-0.7.1 -- /Users/fritzobermeyer/miniconda2/envs/pyro3/bin/python
cachedir: .pytest_cache
benchmark: 3.1.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/fritzobermeyer/github/uber/pyro, inifile: setup.cfg
plugins: xdist-1.20.1, forked-0.2, benchmark-3.1.1, nbval-0.9.1
collected 75 items / 74 deselected

tests/ops/einsum/test_paths.py::test_contract[bd,db,eac->ace] D 	 bd,db,eac->ace
D 	 Pyro path took 0.0004990070010535419s:
         Naive scaling:  5
     Optimized scaling:  3
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  6.000e+01
   Theoretical speedup:  7.200
  Largest intermediate:  2.400e+01 elements
D 	 opt_einsum took 0.0001672200160101056s:
         Naive scaling:  5
     Optimized scaling:  3
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  3.600e+01
   Theoretical speedup:  12.000
  Largest intermediate:  2.400e+01 elements
FAILED
```

**Comment by fritzo** [comment]
Confirmed that it is the `opt_einsum` flop counting that is nondeterministic. We're producing identical paths but getting different costs:
```
DEBUG path = [(0, 1), (1, 0), (0,)]
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  6.000e+01

DEBUG path = [(0, 1), (1, 0), (0,)]
      Naive FLOP count:  4.320e+02
  Optimized FLOP count:  5.000e+01
```

**Comment by neerajprad** [comment]
Thanks for creating the issue, @fritzo. Let us track this externally in https://github.com/dgasmith/opt_einsum/issues/62 itself.

**Comment by fritzo** [comment]
My mistake, this was due to nondeterminism in a test data generator. I'll add the fix to #1430 
```diff
diff --git a/tests/ops/einsum/test_paths.py b/tests/ops/einsum/test_paths.py
index fe1021ca..0ee16890 100644
--- a/tests/ops/einsum/test_paths.py
+++ b/tests/ops/einsum/test_paths.py
@@ -162,7 +162,7 @@ def _test_path(equation, shapes):
 def test_contract(equation):
     inputs, output = equation.split('->')
     inputs = inputs.split(',')
-    symbols = set(output).union(*inputs)
+    symbols = sorted(set(output).union(*inputs))
     sizes = {dim: size for dim, size in zip(symbols, itertools.cycle([2, 3, 4]))}
     shapes = [tuple(sizes[dim] for dim in input_) for input_ in inputs]
     _test_path(equation, shapes)
```
